src/ring_theory/algebra.lean

```diff
@@ -0,0 +1,515 @@
+ring is the under category R ↓ CRing. In the categorical
+setting we have a forgetful functor R-Alg ⥤ R-Mod.
+However here it extends module in order to preserve
+definitional equality in certain cases. -/
+class algebra (R : out_param $ Type u) (A : Type v) [out_param $ comm_ring R] [ring A] extends module R A :=
+(to_fun : R → A) [hom : is_ring_hom to_fun]
+(commutes' : ∀ r x, x * to_fun r = to_fun r * x)
+(smul_def' : ∀ r x, r • x = to_fun r * x)
+
+attribute [instance] algebra.hom
+
+def algebra_map {R : out_param $ Type u} (A : Type v) [out_param $ comm_ring R] [ring A] [algebra R A] (x : R) : A :=
+algebra.to_fun A x
+
+namespace algebra
+
+variables {R : Type u} {S : Type v} {A : Type w}
+variables [comm_ring R] [comm_ring S] [ring A] [algebra R A]
+include R
+
+/-- The codomain of an algebra. -/
+instance : module R A := infer_instance
+instance : has_scalar R A := infer_instance
+
+instance {F : Type u} {K : Type v} [discrete_field F] [ring K] [algebra F K] :
+  vector_space F K :=
+@vector_space.mk F _ _ _ algebra.module
+
+instance : is_ring_hom (algebra_map A : R → A) := algebra.hom A
+
+@[simp] lemma map_add (r s) : algebra_map A (r + s) = algebra_map A r + algebra_map A s :=
+is_ring_hom.map_add _
+
+@[simp] lemma map_zero : algebra_map A 0 = 0 :=
+is_ring_hom.map_zero _
+
+@[simp] lemma map_neg (r) : algebra_map A (-r) = -algebra_map A r :=
+is_ring_hom.map_neg _
+
+@[simp] lemma map_sub (r s) : algebra_map A (r - s) = algebra_map A r - algebra_map A s :=
+is_ring_hom.map_sub _
+
+@[simp] lemma map_mul (r s) : algebra_map A (r * s) = algebra_map A r * algebra_map A s :=
+is_ring_hom.map_mul _
+
+@[simp] lemma map_one : algebra_map A 1 = 1 :=
+is_ring_hom.map_one _
+
+/-- Creating an algebra from a morphism in CRing. -/
+def of_core (i : algebra.core R S) : algebra R S :=
+{ smul := λ c x, i.to_fun c * x,
+  smul_zero := λ x, mul_zero (i.to_fun x),
+  smul_add := λ r x y, mul_add (i.to_fun r) x y,
+  add_smul := λ r s x, show i.to_fun (r + s) * x = _, by rw [i.hom.3, add_mul],
+  mul_smul := λ r s x, show i.to_fun (r * s) * x = _, by rw [i.hom.2, mul_assoc],
+  one_smul := λ x, show i.to_fun 1 * x = _, by rw [i.hom.1, one_mul],
+  zero_smul := λ x, show i.to_fun 0 * x = _, by rw [@@is_ring_hom.map_zero _ _ i.to_fun i.hom, zero_mul],
+  commutes' := λ _ _, mul_comm _ _,
+  smul_def' := λ c x, rfl,
+  .. i }
+
+theorem smul_def (r : R) (x : A) : r • x = algebra_map A r * x :=
+algebra.smul_def' r x
+
+theorem commutes (r : R) (x : A) : x * algebra_map A r = algebra_map A r * x :=
+algebra.commutes' r x
+
+theorem left_comm (r : R) (x y : A) : x * (algebra_map A r * y) = algebra_map A r * (x * y) :=
+by rw [← mul_assoc, commutes, mul_assoc]
+
+@[simp] protected lemma mul_smul (s : R) (x y : A) :
+  x * (s • y) = s • (x * y) :=
+by rw [smul_def, smul_def, left_comm]
+
+@[simp] lemma smul_mul (r : R) (x y : A) :
+  (r • x) * y = r • (x * y) :=
+by rw [smul_def, smul_def, mul_assoc]
+
```

Comment:
`smul_mul_assoc` is a more appropriate name I think

Reply:
done